File: src/main/scala/com/kotori316/fluidtank/tiles/CATTile.java
package com.kotori316.fluidtank.tiles;

import java.util.Collections;
import java.util.Iterator;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Objects;
import java.util.Optional;
import java.util.concurrent.atomic.AtomicLong;
import java.util.stream.Collectors;
import java.util.stream.IntStream;
import java.util.stream.Stream;
import java.util.stream.StreamSupport;

import net.fabricmc.fabric.api.screenhandler.v1.ExtendedScreenHandlerFactory;
import net.fabricmc.fabric.api.transfer.v1.context.ContainerItemContext;
import net.fabricmc.fabric.api.transfer.v1.fluid.FluidStorage;
import net.fabricmc.fabric.api.transfer.v1.fluid.FluidVariant;
import net.fabricmc.fabric.api.transfer.v1.item.InventoryStorage;
import net.fabricmc.fabric.api.transfer.v1.item.ItemStorage;
import net.fabricmc.fabric.api.transfer.v1.storage.Storage;
import net.fabricmc.fabric.api.transfer.v1.storage.StorageView;
import net.fabricmc.fabric.api.transfer.v1.transaction.Transaction;
import net.fabricmc.fabric.api.transfer.v1.transaction.TransactionContext;
import net.minecraft.core.BlockPos;
import net.minecraft.core.Direction;
import net.minecraft.network.FriendlyByteBuf;
import net.minecraft.network.chat.Component;
import net.minecraft.server.level.ServerPlayer;
import net.minecraft.world.MenuProvider;
import net.minecraft.world.entity.player.Inventory;
import net.minecraft.world.entity.player.Player;
import net.minecraft.world.inventory.AbstractContainerMenu;
import net.minecraft.world.level.block.entity.BlockEntity;
import net.minecraft.world.level.block.state.BlockState;
import net.minecraft.world.level.block.state.properties.BlockStateProperties;
import org.jetbrains.annotations.NotNull;
import scala.Option;
import scala.jdk.javaapi.OptionConverters;

import com.kotori316.fluidtank.ModObjects;
import com.kotori316.fluidtank.Utils;
import com.kotori316.fluidtank.fluids.FabricFluidTankStorage;
import com.kotori316.fluidtank.fluids.FluidAction;
import com.kotori316.fluidtank.fluids.FluidAmount;
import com.kotori316.fluidtank.fluids.FluidContainer;
import com.kotori316.fluidtank.fluids.FluidKey;
import com.kotori316.fluidtank.fluids.VariantUtil;

public class CATTile extends BlockEntity implements MenuProvider, ExtendedScreenHandlerFactory {
    // The direction of FACING is facing to you, people expect to use itemBlock targeting an inventory so the chest exists on the opposite side of FACING.
    public List<FluidAmount> fluidCache = Collections.emptyList();

    public CATTile(BlockPos pos, BlockState state) {
        super(ModObjects.CAT_TYPE(), pos, state);
    }

    @SuppressWarnings("UnstableApiUsage")
    public Option<FluidHandlerWrapper> getFluidHandler(Direction direction) {
        assert level != null;
        BlockEntity entity = level.getBlockEntity(getBlockPos().relative(direction));
        if (entity == null) {
            return Option.empty();
        } else {
            var storage = ItemStorage.SIDED.find(level, getBlockPos().relative(direction), direction.getOpposite());
            if (storage instanceof InventoryStorage inventoryStorage && inventoryStorage.supportsInsertion()) {
                return Option.apply(new FluidHandlerWrapper(inventoryStorage));
            } else {
                return Option.empty();
            }
        }
    }

    @Override
    public Component getDisplayName() {
        return getBlockState().getBlock().getName();
    }

    @Override
    public AbstractContainerMenu createMenu(int id, Inventory inv, Player player) {
        return new CATContainer(id, player, getBlockPos());
    }

    @Override
    public void writeScreenOpeningData(ServerPlayer player, FriendlyByteBuf buf) {
        buf.writeBlockPos(getBlockPos());
    }

    @SuppressWarnings("UnstableApiUsage")
    public static class FluidHandlerWrapper implements Storage<FluidVariant>, FluidContainer {
        private final InventoryStorage inventory;

        public FluidHandlerWrapper(InventoryStorage inventory) {
            this.inventory = inventory;
        }

        public Optional<Storage<FluidVariant>> getFluidHandler(int slot) {
            var item = this.inventory.getSlot(slot);
            return Optional.ofNullable(FluidStorage.ITEM.find(item.getResource().toStack(Utils.toInt(item.getAmount())), ContainerItemContext.ofSingleSlot(item)));
        }

        @Override
        public long insert(FluidVariant resource, long maxAmount, TransactionContext transaction) {
            if (resource.isBlank() || maxAmount <= 0) return 0;
            AtomicLong rest = new AtomicLong(maxAmount);
            for (int i = 0; i < this.inventory.getSlots().size(); i++) {
                var handler = getFluidHandler(i);
                handler.map(s -> s.insert(resource, rest.get(), transaction))
                    .ifPresent(l -> rest.addAndGet(-l));
                if (rest.get() <= 0) return maxAmount;
            }
            return maxAmount - rest.get();
        }

        @Override
        public long extract(FluidVariant resource, long maxAmount, TransactionContext transaction) {
            if (resource.isBlank() || maxAmount <= 0) return 0;
            AtomicLong rest = new AtomicLong(maxAmount);
            for (int i = 0; i < this.inventory.getSlots().size(); i++) {
                var handler = getFluidHandler(i);
                handler.map(s -> s.extract(resource, rest.get(), transaction))
                    .ifPresent(l -> rest.addAndGet(-l));
                if (rest.get() <= 0) return maxAmount;
            }
            return maxAmount - rest.get();
        }

        @Override
        public Iterator<StorageView<FluidVariant>> iterator() {
            return inventory.getSlots().stream()
                .map(s -> FluidStorage.ITEM.find(s.getResource().toStack(Utils.toInt(s.getAmount())), ContainerItemContext.ofSingleSlot(s)))
                .filter(Objects::nonNull)
                .flatMap(f -> StreamSupport.stream(f.spliterator(), false))
                .iterator();
        }

        public List<FluidAmount> fluidList() {
            return getFluidAmountStream().toList();
        }

        @NotNull
        public Stream<FluidAmount> getFluidAmountStream() {
            return IntStream.range(0, this.inventory.getSlots().size())
                .mapToObj(this::getFluidHandler)
                .flatMap(Optional::stream)
                .flatMap(s -> StreamSupport.stream(s.spliterator(), false))
                .map(s -> FabricFluidTankStorage.getFluidAmount(s.getResource(), s.getAmount()))
                .collect(Collectors.groupingBy(FluidKey::from, LinkedHashMap::new, Collectors.summingLong(FluidAmount::fabricAmount)))
                .entrySet().stream()
                .map(e -> e.getKey().toAmount(e.getValue()))
                .filter(FluidAmount::nonEmpty);
        }

        @Override
        public FluidAmount fill(FluidAmount resource, FluidAction action) {
            var variant = VariantUtil.convert(resource);
            var amount = resource.fabricAmount();
            try (Transaction transaction = Transaction.openOuter()) {
                var inserted = this.insert(variant, amount, transaction);
                if (action.execute()) {
                    transaction.commit();
                }
                return resource.setAmountF(inserted);
            }
        }

        @Override
        public FluidAmount drain(FluidAmount resource, FluidAction action) {
            var variant = VariantUtil.convert(resource);
            var amount = resource.fabricAmount();
            try (Transaction transaction = Transaction.openOuter()) {
                var extracted = this.extract(variant, amount, transaction);
                if (action.execute()) {
                    transaction.commit();
                }
                return resource.setAmountF(extracted);
            }
        }
    }

    public List<FluidAmount> fluidAmountList() {
        Direction direction = getBlockState().getValue(BlockStateProperties.FACING);
        var opt = OptionConverters.toJava(getFluidHandler(direction));
        return opt.map(FluidHandlerWrapper::fluidList).orElse(Collections.emptyList());
    }

}

File: src/main/scala/com/kotori316/fluidtank/tiles/CATTile.scala
package com.kotori316.fluidtank.tiles

import com.kotori316.fluidtank.ModObjects
import com.kotori316.fluidtank.fluids.{FluidAmount, FluidKey}
import net.minecraft.core.{BlockPos, Direction}
import net.minecraft.network.chat.Component
import net.minecraft.world.MenuProvider
import net.minecraft.world.entity.player.{Inventory, Player}
import net.minecraft.world.level.block.entity.BlockEntity
import net.minecraft.world.level.block.state.BlockState
import net.minecraft.world.level.block.state.properties.BlockStateProperties
import net.minecraftforge.common.capabilities.Capability
import net.minecraftforge.common.util.LazyOptional
import net.minecraftforge.fluids.FluidStack
import net.minecraftforge.fluids.capability.{CapabilityFluidHandler, IFluidHandler, IFluidHandlerItem}
import net.minecraftforge.items.{CapabilityItemHandler, IItemHandlerModifiable}

import scala.collection.immutable
import scala.jdk.OptionConverters.RichOptional

class CATTile(pos: BlockPos, state: BlockState) extends BlockEntity(ModObjects.CAT_TYPE, pos, state) with MenuProvider {
  // The direction of FACING is facing to you, people expect to use itemBlock targeting an inventory so the chest exists on the opposite side of FACING.

  // Client only
  var fluidCache: Seq[FluidAmount] = Seq.empty

  override def getCapability[T](cap: Capability[T], side: Direction): LazyOptional[T] = {
    val direction = getBlockState.getValue(BlockStateProperties.FACING)
    if (side != direction && cap == CapabilityFluidHandler.FLUID_HANDLER_CAPABILITY)
      getFluidHandler(direction).map(t => LazyOptional.of(() => t)).getOrElse(LazyOptional.empty()).cast[T]()
    else
      super.getCapability(cap, side)
  }

  def getFluidHandler(direction: Direction): Option[CATTile.FluidHandlerWrapper] = {
    assert(level != null)
    val entity = level.getBlockEntity(getBlockPos.relative(direction))
    if (entity == null) Option.empty
    else
      entity.getCapability(CapabilityItemHandler.ITEM_HANDLER_CAPABILITY, direction.getOpposite)
        .resolve.toScala
        .collect { case i: IItemHandlerModifiable => new CATTile.FluidHandlerWrapper(i) }
  }

  override def getDisplayName: Component = getBlockState.getBlock.getName

  override def createMenu(id: Int, inv: Inventory, player: Player) = new CATContainer(id, player, getBlockPos)

  def fluidAmountList: Seq[FluidAmount] = {
    val direction = getBlockState.getValue(BlockStateProperties.FACING)
    val opt = getFluidHandler(direction)
    opt.toSeq.flatMap(_.fluidList)
  }
}

object CATTile {
  class FluidHandlerWrapper(val inventory: IItemHandlerModifiable) extends IFluidHandler {
    def getFluidHandler(tank: Int): Option[IFluidHandlerItem] = {
      val stackInSlot = inventory.getStackInSlot(tank)
      if (stackInSlot.isEmpty) Option.empty
      else net.minecraftforge.fluids.FluidUtil.getFluidHandler(stackInSlot).resolve().toScala
    }

    override def getTanks: Int = inventory.getSlots

    override def getFluidInTank(tank: Int): FluidStack = getFluidHandler(tank).map(h => h.drain(Integer.MAX_VALUE, IFluidHandler.FluidAction.SIMULATE)).getOrElse(FluidStack.EMPTY)

    override def getTankCapacity(tank: Int): Int = getFluidHandler(tank).map(h => h.getTankCapacity(0)).getOrElse(0)

    override def isFluidValid(tank: Int, stack: FluidStack): Boolean = getFluidHandler(tank).exists(h => h.isFluidValid(0, stack))

    override def fill(resource: FluidStack, action: IFluidHandler.FluidAction): Int = {
      if (resource == null || resource.isEmpty) return 0
      val toFill = resource.copy
      var i = 0
      while (i < getTanks && !toFill.isEmpty) {
        getFluidHandler(i).foreach { handler =>
          val filled = handler.fill(toFill, action)
          toFill.setAmount(toFill.getAmount - filled)
          if (action.execute())
            inventory.setStackInSlot(i, handler.getContainer)
        }
        i += 1
      }

      resource.getAmount - toFill.getAmount
    }

    override def drain(resource: FluidStack, action: IFluidHandler.FluidAction): FluidStack = {
      if (resource == null || resource.isEmpty) return FluidStack.EMPTY
      val toDrain = resource.copy
      var i = 0
      while (i < getTanks && !toDrain.isEmpty) {
        getFluidHandler(i).foreach { handler =>
          val drained = handler.drain(toDrain, action)
          toDrain.setAmount(toDrain.getAmount - drained.getAmount)
          if (action.execute())
            inventory.setStackInSlot(i, handler.getContainer)
        }
        i += 1
      }

      new FluidStack(resource, resource.getAmount - toDrain.getAmount)
    }

    override def drain(maxDrain: Int, action: IFluidHandler.FluidAction): FluidStack = {
      if (maxDrain <= 0) {
        FluidStack.EMPTY
      } else {
        val first = getFluidAmountStream.map(_.toStack).headOption
        first.map(s => this.drain(new FluidStack(s, Math.min(maxDrain, s.getAmount)), action)).getOrElse(FluidStack.EMPTY)
      }
    }

    def fluidList: Seq[FluidAmount] = getFluidAmountStream.toSeq

    def getFluidAmountStream: immutable.Iterable[FluidAmount] = {
      (0 until this.getTanks).map(this.getFluidInTank)
        .groupMapReduce(FluidKey.from)(_.getAmount.toLong)(_ + _)
        .map { case (key, l) => key toAmount l }
        .filter(_.nonEmpty)
    }
  }
}
